### ISSUE TITLE
initialize conda environment by default during interactive install

### ISSUE THREAD

**Issue by eejd** [conda-forge/miniforge#127]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:
I did a new installation of miniforge (arm64, Apple Silicon) in /opt/miniforge3 (via sudo). It modified the .bash_profile, as bash was used to run the install script. I then ran '/opt/miniforge3/bin/conda init' from the normal user (in the normal user zsh shell) and the init script didn't detect or update the zsh environment files.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list

```
</details>

```
Details about  <code>conda 4.9.2</code> and system ( <code>
     active environment : None
       user config file : /Users/xxx/.condarc
 populated config files : /opt/miniforge3/.condarc
          conda version : 4.9.2
    conda-build version : not installed
         python version : 3.9.2.final.0
       virtual packages : __osx=11.2.1=0
                          __unix=0=0
                          __archspec=1=arm64
       base environment : /opt/miniforge3  (read only)
           channel URLs : https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /opt/miniforge3/pkgs
                          /Users/xxx/.conda/pkgs
       envs directories : /Users/xxx/.conda/envs
                          /opt/miniforge3/envs
               platform : osx-arm64
             user-agent : conda/4.9.2 requests/2.25.1 CPython/3.9.2 Darwin/20.3.0 OSX/11.2.1
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

<details>

```
$ conda info
# packages in environment at /opt/miniforge3:
#
# Name                    Version                   Build  Channel
brotlipy                  0.7.0           py39h46acfd9_1001    conda-forge
bzip2                     1.0.8                h27ca646_4    conda-forge
ca-certificates           2020.12.5            h4653dfc_0    conda-forge
certifi                   2020.12.5        py39h2804cbe_1    conda-forge
cffi                      1.14.5           py39h702c04f_0    conda-forge
chardet                   4.0.0            py39h2804cbe_1    conda-forge
conda                     4.9.2            py39h2804cbe_0    conda-forge
conda-package-handling    1.7.2            py39h51e6412_0    conda-forge
cryptography              3.4.4            py39h6e07874_0    conda-forge
idna                      2.10               pyh9f0ad1d_0    conda-forge
libcxx                    11.0.1               h168391b_0    conda-forge
libffi                    3.3                  h9f76cd9_2    conda-forge
ncurses                   6.2                  h9aa5885_4    conda-forge
openssl                   1.1.1j               h27ca646_0    conda-forge
pip                       21.0.1             pyhd8ed1ab_0    conda-forge
pycosat                   0.6.3           py39h46acfd9_1006    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py39h2804cbe_3    conda-forge
python                    3.9.2           hcbd9b3a_0_cpython    conda-forge
python_abi                3.9                      1_cp39    conda-forge
readline                  8.0                  hc8eb9b7_2    conda-forge
requests                  2.25.1             pyhd3deb0d_0    conda-forge
ruamel_yaml               0.15.80         py39h46acfd9_1004    conda-forge
setuptools                49.6.0           py39h2804cbe_3    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sqlite                    3.34.0               h6d56c25_0    conda-forge
tk                        8.6.10               hf7e6567_1    conda-forge
tqdm                      4.57.0             pyhd8ed1ab_0    conda-forge
tzdata                    2021a                he74cb21_0    conda-forge
urllib3                   1.26.3             pyhd8ed1ab_0    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
yaml                      0.2.5                h642e427_0    conda-forge
zlib                      1.2.11            h31e879b_1009    conda-forge

```
</details>

**Comment by dadances** [comment]
same issuse

**Comment by hmaarrfk** [comment]
I think generally, installing miniforge through sudo is illadvased. We havent explicitely stated this since I think it is old knowledge that i've retained from using conda and reading through Anaconda's documentation.

Can you explain why you are currently installing via sudo? Is there a special usecase we are missing?

**Comment by mwip** [comment]
A previous installation of miniconda3 modified both my `.bashrc` and `.zshrc`. Miniconda only modified my `.bashrc`. 

For now, copying the initialized part from `.bashrc` to `.zshrc` manually fixes it, but it would be nice to see automatic initialization, I guess.

**Comment by mwip** [comment]
To add to yesterday's comment, the user on the system mentioned in the above comment had `bash` as their default shell. 
Now on a different system with `zsh` as the default shell, I found that only the `.zshrc` was modified but not the `.bashrc`. 

This leads me to believe, that only the default shell's config will be affected by `init`. This may lead to cofusions down the line and could be avoided by modifying all configs.

**Comment by hmaarrfk** [comment]
automatically modifying configs is tricky.

At the end, the `conda init` command tells you what got modified.

I think generally, people hit this, and seem to be learning to read the output of the terminal once they get a hand on their tools.

**Comment by jeffheaton** [comment]
My students run into this alot when they are getting setup for the semester.  It would be nice if miniforge could modify the correct shell rc file. This seems to be particularly common on Macs.

**Comment by hmaarrfk** [comment]
The challenge is that we don't try to do anything special here.

we use conda and constructor directly from conda 

do you see the same issue with miniconda?

**Comment by ericpre** [comment]
The `initialize_by_default` issue should be fixed by https://github.com/conda/constructor/pull/458.

There are two things that doesn't make the installer to run `conda init` by default:
1. `initialize_by_default` option was broken for the shell installer (regression introduced in https://github.com/conda/constructor/pull/421)
2. miniforge doesn't set `initialize_by_default=True` in [construct.yaml](https://github.com/conda-forge/miniforge/blob/master/Miniforge3/construct.yaml)

At the same time and to be fair, when running the shell installer, the question asking about running `conda init` is clear, correct and working fine. Using `initialize_by_default=True` in miniforge is a different question.

It would be good to rename the title of this issue to reflect what the issue is about, maybe something along the line of "`conda init` doesn't run by default in shell installer".

**Comment by hmaarrfk** [comment]
@ericpre are you saying this issue is not constrained to OSX?

**Comment by hmaarrfk** [comment]
Yes, this seems to happen for people on linux as well since the default is not to run conda-init
![image](https://user-images.githubusercontent.com/90008/132959418-7ebff513-ba7e-462d-9259-408f89e83890.png)

**Comment by hmaarrfk** [comment]
I think the reason is that constructor doesn't support nested if statements.

I've proposed an alternative fix 
https://github.com/conda/constructor/pull/459

**Comment by lndoja** [comment]
> A previous installation of miniconda3 modified both my `.bashrc` and `.zshrc`. Miniforge only modified my `.bashrc`.
> 
> For now, copying the initialized part from `.bashrc` to `.zshrc` manually fixes it, but it would be nice to see automatic initialization, I guess.

How can I practically to this tho? I'm having the same issue but I'm not too familiar with these kinds of files.

**Comment by hmaarrfk** [comment]
Does this work?
```
~/mambaforge/bin/conda init zsh
~/mambaforge/bin/conda init bash
```

**Comment by matrs** [comment]
2024 and the issue with `zsh` still exists. I'm in manjaro linux, which has zsh by default, and the latest minifoge release only initilized bash.  I had to do `~/mambaforge/bin/conda init zsh`

**Comment by jaimergp** [comment]
The only way we can fix it here is via a `post_install.sh` script in the installer. Otherwise, it'll need a PR in constructor.

**Comment by hmaarrfk** [comment]
I'm not sure it is possible to do this.

Users would have to "source" the script, which not really recommended in terms of best practices.

It somewhat typical for development software to ask users to run an initialization command when a new entry to the PATH is added.
